### PR TITLE
Add tests for blacklisting reactor/agent.

### PR DESCRIPTION
**Pull request by clokep**
Related to #9513, this:

* Adds tests for the blacklisting agent and reactor (we didn't seem to have any?)
* Fixes some type hint errors related to those classes.

I broke it into a few commits that should be standalone, hopefully they make sense!

### CODE REVIEW

**Comment by anoadragon453** [comment]
The error that seems to plague the SyTest logs:

```
2021-03-08 20:46:53,620 - synapse.http.federation.matrix_federation_agent - 289 - INFO - POST-3914 - Failed to connect to localhost:8836: An error occurred while connecting: 99: Cannot assign requested address.
2021-03-08 20:46:53,621 - synapse.http.matrixfederationclient - 552 - INFO - POST-3914 - {PUT-O-30} [localhost:8836] Request failed: PUT matrix://localhost:8836/_matrix/federation/v2/invite/%21NSmQXRsxOFnooAPutw%3Alocalhost%3A8800/%241615236413224mOQKH%3Alocalhost%3A8800: ConnectError('Cannot assign requested address',)
2021-03-08 20:46:54,064 - synapse.metrics - 577 - INFO - None - Collecting gc 1
2021-03-08 20:46:54,668 - synapse.http.federation.matrix_federation_agent - 289 - INFO - POST-3914 - Failed to connect to localhost:8836: An error occurred while connecting: 99: Cannot assign requested address.
2021-03-08 20:46:54,669 - synapse.http.matrixfederationclient - 552 - INFO - POST-3914 - {PUT-O-30} [localhost:8836] Request failed: PUT matrix://localhost:8836/_matrix/federation/v2/invite/%21NSmQXRsxOFnooAPutw%3Alocalhost%3A8800/%241615236413224mOQKH%3Alocalhost%3A8800: ConnectError('Cannot assign requested address',)
2021-03-08 20:46:56,430 - synapse.http.federation.matrix_federation_agent - 289 - INFO - POST-3914 - Failed to connect to localhost:8836: An error occurred while connecting: 99: Cannot assign requested address.
2021-03-08 20:46:56,431 - synapse.http.matrixfederationclient - 552 - INFO - POST-3914 - {PUT-O-30} [localhost:8836] Request failed: PUT matrix://localhost:8836/_matrix/federation/v2/invite/%21NSmQXRsxOFnooAPutw%3Alocalhost%3A8800/%241615236413224mOQKH%3Alocalhost%3A8800: ConnectError('Cannot assign requested address',)
2021-03-08 20:46:58,732 - synapse.http.federation.matrix_federation_agent - 289 - INFO - POST-3914 - Failed to connect to localhost:8836: An error occurred while connecting: 99: Cannot assign requested address.
2021-03-08 20:46:58,733 - synapse.http.matrixfederationclient - 552 - INFO - POST-3914 - {PUT-O-30} [localhost:8836] Request failed: PUT matrix://localhost:8836/_matrix/federation/v2/invite/%21NSmQXRsxOFnooAPutw%3Alocalhost%3A8800/%241615236413224mOQKH%3Alocalhost%3A8800: ConnectError('Cannot assign requested address',)
2021-03-08 20:46:58,735 - synapse.http.server - 98 - ERROR - POST-3914 - Failed handle request via 'RoomCreateRestServlet': <SynapseRequest at 0x7fceebb54da0 method='POST' uri='/_matrix/client/r0/createRoom?access_token=<redacted>' clientproto='HTTP/1.1' site='8800'>
Traceback (most recent call last):
  File "/venv/lib/python3.5/site-packages/synapse/http/matrixfederationclient.py", line 481, in _send_request
    response = await request_deferred
  File "/venv/lib/python3.5/site-packages/twisted/internet/defer.py", line 1416, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/venv/lib/python3.5/site-packages/twisted/python/failure.py", line 512, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/venv/lib/python3.5/site-packages/synapse/util/patch_inline_callbacks.py", line 142, in check_yield_points_inner
    d = result.throwExceptionIntoGenerator(gen)
  File "/venv/lib/python3.5/site-packages/twisted/python/failure.py", line 512, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/venv/lib/python3.5/site-packages/synapse/http/federation/matrix_federation_agent.py", line 191, in request
    self._agent.request(method, uri, request_headers, bodyProducer)
  File "/venv/lib/python3.5/site-packages/synapse/util/patch_inline_callbacks.py", line 190, in check_yield_points_inner
    result = yield d
  File "/venv/lib/python3.5/site-packages/twisted/internet/defer.py", line 1416, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/venv/lib/python3.5/site-packages/twisted/python/failure.py", line 512, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/venv/lib/python3.5/site-packages/synapse/http/federation/matrix_federation_agent.py", line 296, in _do_connect
    raise first_exception
  File "/venv/lib/python3.5/site-packages/synapse/http/federation/matrix_federation_agent.py", line 283, in _do_connect
    endpoint.connect(protocol_factory)
twisted.internet.error.ConnectError: An error occurred while connecting: 99: Cannot assign requested address.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/venv/lib/python3.5/site-packages/synapse/http/server.py", line 259, in _async_render_wrapper
    callback_return = await self._async_render(request)
  File "/venv/lib/python3.5/site-packages/synapse/http/server.py", line 447, in _async_render
    callback_return = await raw_callback_return
  File "/venv/lib/python3.5/site-packages/synapse/rest/client/v1/room.py", line 83, in on_POST
    requester, self.get_room_config(request)
  File "/venv/lib/python3.5/site-packages/synapse/handlers/room.py", line 818, in create_room
    content=content,
  File "/venv/lib/python3.5/site-packages/synapse/handlers/room_member.py", line 621, in update_membership_locked
    require_consent=require_consent,
  File "/venv/lib/python3.5/site-packages/synapse/handlers/room_member.py", line 248, in _local_membership_update
    ratelimit=ratelimit,
  File "/venv/lib/python3.5/site-packages/synapse/util/metrics.py", line 92, in measured_func
    r = await func(self, *args, **kwargs)
  File "/venv/lib/python3.5/site-packages/synapse/handlers/message.py", line 989, in handle_new_client_event
    requester, event, context, ratelimit=ratelimit, extra_users=extra_users
  File "/venv/lib/python3.5/site-packages/synapse/handlers/message.py", line 1177, in persist_and_notify_client_event
    invitee.domain, event
  File "/venv/lib/python3.5/site-packages/synapse/handlers/federation.py", line 1312, in send_invite
    pdu=event,
  File "/venv/lib/python3.5/site-packages/synapse/federation/federation_client.py", line 761, in send_invite
    content = await self._do_send_invite(destination, pdu, room_version)
  File "/venv/lib/python3.5/site-packages/synapse/federation/federation_client.py", line 795, in _do_send_invite
    "invite_room_state": pdu.unsigned.get("invite_room_state", []),
  File "/venv/lib/python3.5/site-packages/synapse/federation/transport/client.py", line 311, in send_invite_v2
    destination=destination, path=path, data=content, ignore_backoff=True
  File "/venv/lib/python3.5/site-packages/synapse/http/matrixfederationclient.py", line 720, in put_json
    timeout=timeout,
  File "/venv/lib/python3.5/site-packages/synapse/http/matrixfederationclient.py", line 297, in _send_request_with_optional_trailing_slash
    response = await self._send_request(request, **send_request_args)
  File "/venv/lib/python3.5/site-packages/synapse/http/matrixfederationclient.py", line 485, in _send_request
    raise RequestSendFailed(e, can_retry=True) from e
synapse.api.errors.RequestSendFailed: Failed to send request: ConnectError: An error occurred while connecting: 99: Cannot assign requested address.
```

File: synapse/http/client.py [hunk 1]
```diff
@@ -176,14 +178,14 @@ def _callback() -> None:
             # valid results.
             if not has_bad_ip:
                 for i in addresses:
-                    r.addressResolved(i)
-            r.resolutionComplete()
+                    recv.addressResolved(address)
+            recv.resolutionComplete()
```
Comment (anoadragon453) [inline synapse/http/client.py:182]: Do these `recv`s need to be `recv()`?

Comment (clokep) [reply synapse/http/client.py:182]: I mentioned this in 315061fcb9362a9bc4ee0f33dd05e370b5f99332, but `recv` does *not* need to be instantiated. (It actually shouldn't be, it isn't a callable necessarily!) The only reason the current code works is that the [implementation of `EndpointReceiver`](https://github.com/twisted/twisted/blob/release-21.2.0-10091/src/twisted/internet/endpoints.py#L943-L955) happens to use a class with all static methods, so instantiating it essentially has the same methods called as if it isn't instantiated.

tl;dr `recv` is meant to be an instance, not a class.

**Comment by clokep** [comment]
I had a typo in one of the commits which was causing the sytest failure. This should be ready for review now!

File: tests/http/test_client.py [hunk 2]
```diff
@@ -119,3 +130,114 @@ def test_content_length(self):
...
+            ip_whitelist=self.ip_whitelist,
+            ip_blacklist=self.ip_blacklist,
+        )
+
+        # The unsafe IPs should be rejected.
+        self.failureResultOf(
+            agent.request(b"GET", b"http://" + self.unsafe_ip), SynapseError
+        )
+
+        # The safe and unsafe domains and safe IPs should be accepted.
```
Comment (anoadragon453) [inline tests/http/test_client.py:214]: Why is the unsafe domain accepted? Shouldn't it resolve to the unsafe IP and then be blocked?

**Review by anoadragon453 — COMMENTED** [review]
lgtm other than this bit that I don't quite understand.

File: tests/http/test_client.py (hunk 2 above)
Comment (clokep) [reply tests/http/test_client.py:214]: The agent doesn't do any resolution, it only attempts to block access when IPs are directly accessed. The reactor ensures that resolved IPs can be blocked effectively. This is tested above. You should use both of them in tandem to ensure all cases are blocked.

Comment (anoadragon453) [reply tests/http/test_client.py:214]: I see, and the docstring of `BlacklistingAgentWrapper` does indeed say as much. Ah well, thank you!

**Review by anoadragon453 — APPROVED** [review]
lgtm!
